exploration/data-driven-tests.md

```diff
@@ -0,0 +1,243 @@
+
+The README specifies that test cases and expected results are to be located in separate files (including the rationale for this).
+
+#### Test file example
+
+`example_1_test.json`
+
+```jsonc
+{
+  "Test scenario": "example_1",
+  "description": "Test cases for XYZ",
+  "testType": "syntax", // Tests will require different setup steps or function calls depending on their purpose.
+  "tests": [
+    {
+      "label": "0000",
+      "locale": "en-US",
+      "pattern": "{Some MF2 pattern}",
+      "options": {}, // Optional configuration
+      "input": { "namedArg": "foo" } // Arguments to the function being tested, such as a message.formatToString() function. May vary with testType.
+    }
+    // ...
+  ]
+}
+```
+
+#### Verification file example
+
+`example_1_verify.json`
+
+```jsonc
+{
+  "Test scenario": "example_1",
+  "verifications": [
+    {
+      "label": "0000",
+      "verify": "Expected result"
+    }
+    // ...
+  ]
+}
+```
+
+### YAML-to-JSON test transpilation (optional)
+
+JSON does not support multiline strings. Test files may need to include `\n` line breaks in order to capture multiline patterns, which could cause readability issues. YAML syntax offers improved readability in this situation.
+
+We can include a script that converts YAML-formatted "test" and "verify" files to JSON. Assuming both the YAML- and JSON-format tests are committed, the JSON can remain the single source of truth for the tests, which aligns with the expectations of Unicode Data Driven Tests.
+
+### Test content
+
+#### Syntax tests
+
+These tests evaluate the pattern based on the runtime arguments. Formatters are shown as stringified representations of the function because formatter output is tested separately.
+
+Example:
+
+```jsonc
+{
+  "label": "Renders multiple inputs in formatted string",
+  "locale": "en-US",
+  "pattern": "{{$strArg :string} and {$numArg :number}}",
+  "inputs": {
+    "strArg": { "type": "string", "value": "foo" },
+    "numArg": { "type": "number", "value": 123 }
+  }
+  // "verify":  "{ formatter: "string", value: "foo" } and { formatter: number, value: 123 }"
+}
+```
+
+#### Formatter tests
+
+These tests focus on the standard registry's formatters (e.g. `:number`, `:datetime`). They cover the different options that can be passed to each formatter (e.g. `offset`, `skeleton`).
+
+Example:
+
+```jsonc
+{
+  "label": "Skeleton affects datetime format",
+  "locale": "en-US",
+  "pattern": "{$givenDateTime :datetime skeleton=yMMMdE}",
+  "inputs": {
+    "givenDateTime": { "type": "datetime", "value": "2000-12-31T00:00:00.000Z" }
+  }
+  // "verify":  "Sun, 31 Dec 2000"
+}
+```
+
+#### Selector tests
+
+These are extensive tests of the cases within a `match` statement. Testing of multiple selectors is included.
+
+Example:
+
+```jsonc
+{
+  "label": "Matches numbers other than one",
+  "locale": "en-US",
+  "pattern": "match {$arg :number} when one {one} when * {*}",
```

Comment:
Yes, the output string value are matching purely as a convention. Now you've mentioned it, I think this approach could be misleading - I'll change them to avoid the need for more explanation. I'll also add a multi-selector example 👍